File: training/3.1/stamps/stamps.cpp
/*
ID: zhbngch1
PROG: stamps
LANG: C++
*/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <utility>
#include <algorithm>
#include <limits>

using namespace std;

#define ARRAYSIZE 2000020

int main() {
    int results[ARRAYSIZE];
    for (int i = 0; i < ARRAYSIZE; i++)
    {
        results[i] = 0;
    }

    ofstream fout ("stamps.out");
    ifstream fin ("stamps.in");
    int K, N;
    fin >> K;
    fin >> N;
    vector<int> numbers;
    for (int i = 0; i < N; i++)
    {
        int val;
        fin >> val;
        numbers.push_back(val);
    }

    sort(numbers.begin(), numbers.end(), greater<int>());
    /* for (auto number: numbers)
    {
        cout << number << ' ';
    }
    cout << endl;*/

    for (int i = 1; ; i++)
    {
        for (auto number : numbers)
        {
            if (i - number >= 0)
            {
                if (results[i] == 0 || results[i] > results[i-number] + 1)
                {
                    results[i] = results[i-number] + 1;
                }
            }
        }
        if (results[i] == 0 || results[i] > K)
        {
            fout << i-1 << endl;
            return 0;
        }
    }

    return 0;
}

File: training/3.1/stamps/stamps.py
"""
PROG: stamps
LANG: PYTHON3

"""
import sys
with open('stamps.in', 'r') as fin:
  lines = fin.readlines()

K, N = map(int, lines[0].split())

bases = []
for line in lines[1:]:
  bases += list(map(int, line.split()))

bases.sort(reverse = True)
print(bases)

listSize = 1000000
results = [sys.maxsize] * listSize
results[0] = 0

result = 0
for i in range(1, listSize):
  found = False
  count = 0
  for base in bases:
    if i - base >= 0 and results[i] > results[i - base] + 1:
      results[i] = results[i - base] + 1
      count += 1
    if count > 2:
      break
  if results[i] > K or results[i] == sys.maxsize:
        found = True
  if found == True:
    result = i - 1
    break

with open('stamps.out', 'w') as fout:
  fout.write(str(result) + '\n')